services/fission-rmq-connector/main.go

```diff
@@ -0,0 +1,219 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rabbitmq/amqp091-go"
+
+	"sensorbucket.nl/sensorbucket/internal/env"
+	"sensorbucket.nl/sensorbucket/pkg/mq"
+)
+
+var (
+	opsProcessed = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "sensorbucket_fissionrmqconnect_processed",
+		Help: "total number of message queue messages received",
+	})
+	opsSuccesful = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "sensorbucket_fissionrmqconnect_success",
+		Help: "total number of message queue messages succesfully processed",
+	})
+	opsFails = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "sensorbucket_fissionrmqconnect_fail",
+		Help: "total number of message queue messages failed to process",
+	})
+)
+
+func main() {
+	if err := Run(); err != nil {
+		panic(err)
+	}
+}
+
+var (
+	AMQP_HOST     = env.Must("HOST")
+	AMQP_QUEUE    = env.Must("QUEUE_NAME")
+	AMQP_PREFETCH = env.Could("AMQP_PREFETCH", "5")
+	AMQP_XCHG     = env.Must("EXCHANGE")
+	HTTP_ENDPOINT = env.Must("HTTP_ENDPOINT")
+	MAX_RETRIES   = env.Could("MAX_RETRIES", "3")
+	METRICS_ADDR  = env.Could("METRICS_ADDR", ":2112")
+)
+
+func Run() error {
+	prefetch, err := strconv.Atoi(AMQP_PREFETCH)
+	if err != nil {
+		return err
+	}
+	maxRetries, err := strconv.Atoi(MAX_RETRIES)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Consuming from queue: %s and producing to exchange: %s\n", AMQP_QUEUE, AMQP_XCHG)
+	conn := mq.NewConnection(AMQP_HOST)
+	successChan := conn.Publisher(AMQP_XCHG, func(c *amqp091.Channel) error {
+		return nil
+	})
+	consumeChan := conn.Consume(AMQP_QUEUE, func(c *amqp091.Channel) error {
+		// Cannot set autodelete because KEDA can't scale if the queue does not exist at all
+		_, err := c.QueueDeclare(AMQP_QUEUE, true, false, false, false, nil)
+		if err != nil {
+			return err
+		}
+		err = c.ExchangeDeclare(AMQP_XCHG, "topic", true, false, false, false, nil)
+		if err != nil {
+			return err
+		}
+		err = c.QueueBind(AMQP_QUEUE, AMQP_QUEUE, AMQP_XCHG, false, nil)
+		if err != nil {
+			return err
+		}
+		err = c.Qos(prefetch, 0, true)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	go conn.Start()
+
+	connector := Connector{
+		Name:       fmt.Sprintf("%s-(%s)", os.Getenv("SOURCE_NAME"), os.Getenv("HOSTNAME")),
+		Endpoint:   HTTP_ENDPOINT,
+		MaxRetries: maxRetries,
+		Result:     successChan,
+	}
+
+	concurrency := make(chan int, prefetch)
+	go func(incoming <-chan amqp091.Delivery) {
+		for delivery := range incoming {
+			concurrency <- 1
+			go func(delivery amqp091.Delivery) {
+				log.Printf("Processing: %s\n", delivery.MessageId)
+				connector.handleDelivery(delivery)
+				<-concurrency
```

Comment:
what does this concurrency mechanism do? something with batching? However I don't see any changes in the chan. it just keeps getting -1?

Reply:
The channel will block if there are more than `prefetch` amount of messages in flight. It is a bit redundant as we also set the prefetch in AMQP, as such AMQP should not provide more than `prefetch` amount of messages.